File: Fixed/src/gen_exp.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "fixed.h"

int main(int argc, char **argv) {
    int i, j;

    FILE *f = fopen("exp.h", "w");

    if (f == NULL) return EXIT_FAILURE;

    fprintf(f, "static fixed ln_2 = %i;\n", fixed_double(log(2)));

    fprintf(f, "static fixed ln_table[] = {\n");
    j = 0;
    for (i = 0x0000;i <= 0xFF00;i += 0x0100) {
        if (i != 0) fprintf(f, "%i", fixed_double(log(fixed_todouble(i))));
        else fprintf(f, "%i", -1);

        if (i != 0xFF00) fprintf(f, ", ");
        else if (j != 15) fprintf(f, "\n");

        j++;
        if (j == 16) {
            fprintf(f, "\n");
            j = 0;
        }
    }
    fprintf(f, "};\n\n");

    fprintf(f, "static fixed exp_table[] = {\n");
    j = 0;
    for (i = 0;i <= 30;i++) {
        fixed r = fixed_double(exp(pow(2, i - 16)));
        if (r < 0) r = 0;
        fprintf(f, "%i", r);

        if (i != 30) fprintf(f, ", ");
        else if (j != 15) fprintf(f, "\n");

        j++;
        if (j == 16) {
            fprintf(f, "\n");
            j = 0;
        }
    }
    fprintf(f, "};\n");

    fclose(f);

    return EXIT_SUCCESS;
}

File: Fixed/src/gen_exp.py
import math

FIXED_SHIFT = 16

def fixed(d):
    return math.ceil(d * (1 << FIXED_SHIFT))

print("static fixed ln_2 = {};".format(fixed(math.log(2))))

print("static fixed ln_table[] = {")

j = 0
for i in range(0x0000, 0x10000, 0x100):
    if j != 15:
        j += 1
        end = ', '
    else:
        j = 0
        end = ',\n' if i != 0x10000 - 0x100 else '\n'
    if i != 0:
        print(fixed(math.log(i / (1 << FIXED_SHIFT))), end=end)
    else:
        print("-1", end=end)

print("};")
print()
print("static fixed exp_table[] = {")

for i in range(0, 31):
    r = fixed(math.exp(math.pow(2, i - 16))) if i < 20 else 0

    if i == 15:
        end = ',\n'
    elif i == 30:
        end = '\n'
    else:
        end = ', '

    print(r, end=end)

print("};")
